driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -0,0 +1,82 @@
+use super::*;
+use crate::contracts::stablex_contract::batch_exchange;
+use anyhow::{anyhow, bail, Result};
+use block_timestamp::BlockTimestamp;
+use ethcontract::{contract::Event, errors::ExecutionError};
+use futures::{
+    channel::oneshot,
+    future::{BoxFuture, FutureExt},
+    select,
+    stream::{BoxStream, StreamExt as _},
+};
+use orderbook::Orderbook;
+use std::sync::{Arc, Mutex};
+
+// TODO:
+// `pub struct UpdatingOrderbook` that also owns an `Arc<Mutex<Orderbook>>` and creates a
+// background thread running `UpdatingOrderbookThread::update_with_events`.
+// This way `UpdatingOrderbookThread` is independently testable.
+
+/// Update the orderbook with events from the stream forever or until exit_indicator is dropped.
+///
+/// Returns Ok when exit_indicator is dropped.
+/// Returns Err if the stream ends.
+async fn update_with_events_forever(
+    orderbook: Arc<Mutex<Orderbook>>,
+    mut block_timestamp: impl BlockTimestamp,
+    exit_indicator: oneshot::Receiver<()>,
+    past_events: BoxFuture<'static, Result<Vec<Event<batch_exchange::Event>>, ExecutionError>>,
+    stream: BoxStream<'static, Result<Event<batch_exchange::Event>, ExecutionError>>,
+) -> Result<()> {
+    // `select!` requires the futures to be fused.
+    // By selecting over exit_indicator and the future/stream we make sure to exit immediately when
+    // exit_indicator is dropped without waiting for the future or the stream to produce a value.
+    let mut exit_indicator = exit_indicator.fuse();
+    let mut past_events = past_events.fuse();
+    let mut stream = stream.fuse();
+
+    let past_events = select! {
+        past_events = past_events => past_events?,
+        _ = exit_indicator => return Ok(()),
+    };
+    for event in past_events {
+        handle_event(&orderbook, &mut block_timestamp, event).await?;
+    }
+
+    loop {
+        let event = select! {
+                event = stream.next() =>
+                    event.ok_or(anyhow!("stream ended"))??,
+                _ = exit_indicator => return Ok(()),
+        };
+        handle_event(&orderbook, &mut block_timestamp, event).await?;
```

Comment:
Thinking about this a bit more last night, you will probably have to select on all 3 in the loop (since `past_events` is fused it will only resolve once :smile:). That way the stream actually gets polled and `eth_newEventFilter` gets called so the filter gets setup on the node and we don't have a race condition where events come in between the past events being done and the stream getting polled for the first time.

Reply:
I have changed this. Technically it still doesn't guarantee that `stream` gets polled first. Not sure if we can improve this. Maybe manually call poll once on the stream?

Reply:
This will only be an issue if the any of the futures resolve immediately which will never happen in our case.